### ISSUE TITLE
[REQUEST]  How to Subclass the trainer and override the compute loss function  for custom behavior?

### ISSUE THREAD

**Issue by zhangsanfeng86** [microsoft/DeepSpeed#3079]
**Is your feature request related to a problem? Please describe.**

Hi, 
Thanks for your great job!

In the Trainer of HuggleFace, I can subclass thr trainer, and override the compute_loss function for custom behavior. As follow:
https://github.com/huggingface/transformers/blob/198c335d219a5eb4d3f124fdd1ce1a9cd9f78a9b/src/transformers/trainer.py#L2006

**Describe the solution you'd like**

In Deepspeed, is it possible to subclass thr trainer, and override the compute_loss function for custom behavior?
If True, Please tell how to do!

Thank you !
**Describe alternatives you've considered**

**Additional context**

**Comment by zhangsanfeng86** [comment]
@stas00  Could you give me some help?

Thank you

**Comment by stas00** [comment]
I don't quite understand what you're asking, but if it's HF Trainer-related you're asking in the wrong repo, @zhangsanfeng86 - please file your request at https://github.com/huggingface/transformers/issues instead.

And when you file it please explain better your query as I don't understand what you're asking.

In general you can use HF Trainer with deepspeed and shouldn't be a problem subclassing it, as long as you don't override deepspeed-specific integration methods.

It'd be better if you have shown what you have tried and what difficulties have you run into including an easy way to reproduce any problems.

**Comment by zhangsanfeng86** [comment]
@stas00  Thanks for your helpful reply.

The trainer in huggingface is very helpful.  With the help of hugging face, deepspeed and zero_optimization, it is possible to finetune the model of bigscience/bloom-7b1 on A100 (40GB).

But for facebook/opt-30b, when finetune, even using zero_optim stage3, the program will kill because of GPU OOM.
Can hugglingface's trainer split large models into multiple GPUs?

As far as I know, The megatron-deepspeed can loads large models on multiple GPUs, but when using deepspeed, I don't know how to subclass and override the compute loss function.

**Comment by stas00** [comment]
> Can hugglingface's trainer split large models into multiple GPUs?

This is done for you automatically in HF Trainer via Deepspeed integration. Please read: https://huggingface.co/docs/transformers/main/main_classes/deepspeed#deepspeed-trainer-integration

You can estimate memory requirements here: https://deepspeed.readthedocs.io/en/latest/memory.html and then you will need more memory for activations.

**Comment by zhangsanfeng86** [comment]
Thank you , I'll try!

**Comment by zhangsanfeng86** [comment]
@stas00 
The HF Trainer is not used，
then， how to override the compute_loss function for custom behavior

<img width="935" alt="image" src="https://user-images.githubusercontent.com/73547638/227133552-8be31dd4-55cb-413d-9be9-c986d7fe739b.png">

**Comment by stas00** [comment]
Sure, you can do your own Deepspeed integration. But this is far from trivial. 

The code you snapshot'ed is the only requirement to get `zero.Init` working when you do `from_pretrained` - you will have to implement everything else yourself.

`compute_loss` has nothing to do with the Deepspeed integration.

The easiest path is to subclass HF Trainer where everything is already done for you. 

You also have many examples that do specific things https://github.com/huggingface/transformers/tree/main/examples/pytorch based on HF Trainer.

**Comment by zhangsanfeng86** [comment]
OK , thank you for your prompt response. @stas00
